datafusion/sqllogictest/test_files/join_disable_repartition_joins.slt

```diff
@@ -72,11 +72,11 @@ SELECT t1.a, t1.b, t1.c, t2.a as a2
  ON t1.d = t2.d ORDER BY a2, t2.b
  LIMIT 5
 ----
-0 0 0 0
-0 0 2 0
-0 0 3 0
-0 0 6 0
-0 0 20 0
+1 3 95 0
```

Comment:
why did this output change? is it because the order on t2.a is a tie?

Reply:
Well, yes -- current behaviour for indices-mathching function is to [iterate](https://github.com/apache/arrow-datafusion/blob/c2e768052c43e4bab6705ee76befc19de383c2cb/datafusion/physical-plan/src/joins/hash_join.rs#L881) over inverted probe-side indices and attach build-side indices to them (HashMap + Vector data structure for also emits build-side indices in reverse order), and after matching whole probe-side side, resulting arrays [inverted](https://github.com/apache/arrow-datafusion/blob/c2e768052c43e4bab6705ee76befc19de383c2cb/datafusion/physical-plan/src/joins/hash_join.rs#L915-L916) again -- it allows to return right-left side indices in their natural order.

In case of partial output -- I can't see any other option besides iterating probe-side naturally (otherwise the order of record would be broken as there is no "full batch" anymore to re-sort it), but in the same time build-side is stored in same data structure with reverse order.

So, it's a side effect -- hash join still maintains probe-side input order, but not for build-side anymore (guess it can potentially be achieved by tweaking `collect_build_side` function)

Reply:
I believe preserving the order of the build side is important for further aggregations. We can avoid unnecessary sorting by leveraging the build-side lexical order. This pattern is common in Datafusion and warrants investing in order preservation.

Reply:
I agree that build-side order is not that important (at least for majority of cases -- anyway it will be affected by probe-side ordering), but on the other side it's reasonable not to break current behaviour.

At first glance, the cost is increase of `join_build_time` caused by switching current `JoinHashMap` building [process](https://github.com/apache/arrow-datafusion/blob/f3c9009e50fdb5811c522bfb07ba29ac04cd1d22/datafusion/physical-plan/src/joins/hash_join.rs#L729) from LIFO (last value stored in map) to FIFO (first value stored in map) -- it'll probably require more traversals over `JoinHashMap.next` to place all the elements, and affect joins with high share of non-unique values on build side.

Reply:
@metesynnada, my bad, didn't notice those special cases in `calculate_join_output_ordering` for inner joins.

I guess, I'll make some research on [this one](https://github.com/apache/arrow-datafusion/pull/6679/files#r1233938719) and ideas related to it.

Thank you!